src/core/mempool.rs

```diff
@@ -1175,71 +1249,137 @@ impl MemPoolDB {
                 break;
             }
 
-            let start_with_no_estimate = remember_start_with_estimate.unwrap_or_else(|| {
-                tx_consideration_sampler.sample(&mut rng) < settings.consider_no_estimate_tx_prob
-            });
+            let start_with_no_estimate =
+                tx_consideration_sampler.sample(&mut rng) < settings.consider_no_estimate_tx_prob;
 
-            match self.get_next_tx_to_consider(start_with_no_estimate)? {
-                ConsiderTransactionResult::NoTransactions => {
-                    debug!("No more transactions to consider in mempool");
-                    break;
+            // First, try to read from the retry list
+            let (candidate, update_estimate) = match candidate_cache.next() {
+                Some(tx) => {
+                    let update_estimate = tx.fee_rate.is_none();
+                    (tx, update_estimate)
                 }
-                ConsiderTransactionResult::UpdateNonces(addresses) => {
-                    // if we need to update the nonce for the considered transaction,
-                    //  use the last value of start_with_no_estimate on the next loop
-                    remember_start_with_estimate = Some(start_with_no_estimate);
-                    let mut last_addr = None;
-                    for address in addresses.into_iter() {
-                        debug!("Update nonce"; "address" => %address);
-                        // do not recheck nonces if the sponsor == origin
-                        if last_addr.as_ref() == Some(&address) {
-                            continue;
+                None => {
+                    // When the retry list is empty, read from the mempool db,
+                    // randomly selecting from either the null fee-rate transactions
+                    // or those with fee-rate estimates.
+                    let opt_tx = if start_with_no_estimate {
+                        null_iterator
+                            .next()
+                            .map_err(|err| Error::SqliteError(err))?
+                    } else {
+                        fee_iterator.next().map_err(|err| Error::SqliteError(err))?
+                    };
+                    match opt_tx {
```

Comment:
Regarding this `match` case, can you add a unit test if there isn't one already to verify that all transactions will be visited once, regardless of what `consider_no_estimate_tx_prob` is?  Thanks!

Reply:
Added in fa95c06d43cec7a3c69a04df558c104756da3ac7